File: src/provider/CloudJsonProvider.php
<?php

namespace pocketcloud\cloud\provider;

use pocketcloud\cloud\cache\impl\InGameModuleCache;
use pocketcloud\cloud\cache\impl\MaintenanceListCache;
use pocketcloud\cloud\cache\impl\NotificationListCache;
use pocketcloud\cloud\config\Config;
use pocketcloud\cloud\group\ServerGroup;
use pocketcloud\cloud\template\Template;
use pocketcloud\cloud\util\promise\Promise;
use const pocketcloud\IN_GAME_PATH;
use const pocketcloud\SERVER_GROUPS_PATH;
use const pocketcloud\TEMPLATES_PATH;

final class CloudJsonProvider extends CloudProvider {

    private Config $templatesConfig;
    private Config $serverGroupsConfig;
    private Config $modulesConfig;
    private Config $notificationsList;
    private Config $maintenanceList;

    public function __construct() {
        $this->templatesConfig = new Config(TEMPLATES_PATH . "templates.json");
        $this->serverGroupsConfig = new Config(SERVER_GROUPS_PATH . "groups.json");
        $this->modulesConfig = new Config(IN_GAME_PATH . "modules.json");
        $this->notificationsList = new Config(IN_GAME_PATH . "notifications.json");
        $this->maintenanceList = new Config(IN_GAME_PATH . "maintenanceList.json");

        foreach ($this->maintenanceList->getAll() as $player => $enabled) if ($enabled) MaintenanceListCache::add($player);
        foreach ($this->notificationsList->getAll() as $player => $enabled) if ($enabled) NotificationListCache::add($player);
        foreach ($this->modulesConfig->getAll() as $module => $enabled) InGameModuleCache::setModuleState($module, $enabled);
    }

    public function addTemplate(Template $template): Promise {
        $this->templatesConfig->set($template->getName(), $template->write());
        return $this->templatesConfig->save() ? Promise::resolved() : Promise::rejected();
    }

    public function removeTemplate(Template $template): Promise {
        $this->templatesConfig->remove($template->getName());
        return $this->templatesConfig->save() ? Promise::resolved() : Promise::rejected();
    }

    public function editTemplate(Template $template, array $newData): Promise {
        $this->templatesConfig->set($template->getName(), $newData);
        return $this->templatesConfig->save() ? Promise::resolved() : Promise::rejected();
    }

    public function getTemplate(string $template): Promise {
        $promise = new Promise();

        $data = $this->templatesConfig->get($template);
        if (($template = Template::read($data)) !== null) {
            $promise->resolve($template);
        } else $promise->reject();

        return $promise;
    }

    public function checkTemplate(string $template): Promise {
        return Promise::resolved($this->templatesConfig->has($template));
    }

    public function getTemplates(): Promise {
        $promise = new Promise();

        $templates = [];
        $data = $this->templatesConfig->getAll();
        foreach ($data as $template) {
            if (($template = Template::read($template)) !== null) $templates[$template->getName()] = $template;
        }

        $promise->resolve($templates);
        return $promise;
    }

    public function addServerGroup(ServerGroup $serverGroup): Promise {
        $this->serverGroupsConfig->set($serverGroup->getName(), $serverGroup->write());
        return $this->serverGroupsConfig->save() ? Promise::resolved() : Promise::rejected();
    }

    public function removeServerGroup(ServerGroup $serverGroup): Promise {
        $this->serverGroupsConfig->remove($serverGroup->getName());
        return $this->serverGroupsConfig->save() ? Promise::resolved() : Promise::rejected();
    }

    public function editServerGroup(ServerGroup $serverGroup, array $newData): Promise {
        $this->serverGroupsConfig->set($serverGroup->getName(), $newData);
        return $this->serverGroupsConfig->save() ? Promise::resolved() : Promise::rejected();
    }

    public function getServerGroup(string $serverGroup): Promise {
        $promise = new Promise();

        $data = $this->serverGroupsConfig->get($serverGroup);
        if (($serverGroup = ServerGroup::read($data)) !== null) {
            $promise->resolve($serverGroup);
        } else $promise->reject();

        return $promise;
    }

    public function checkServerGroup(string $serverGroup): Promise {
        return Promise::resolved($this->serverGroupsConfig->has($serverGroup));
    }

    public function getServerGroups(): Promise {
        $promise = new Promise();

        $serverGroups = [];
        $data = $this->serverGroupsConfig->getAll();
        foreach ($data as $serverGroup) {
            if (($serverGroup = ServerGroup::read($serverGroup)) !== null) $serverGroups[$serverGroup->getName()] = $serverGroup;
        }

        $promise->resolve($serverGroups);
        return $promise;
    }

    public function setModuleState(string $module, bool $enabled): Promise {
        $this->modulesConfig->set($module, $enabled);
        InGameModuleCache::setModuleState($module, $enabled);
        return $this->modulesConfig->save() ? Promise::resolved() : Promise::rejected();
    }

    public function getModuleState(string $module): Promise {
        return Promise::resolved($this->modulesConfig->get($module, false));
    }

    public function enablePlayerNotifications(string $player): Promise {
        $this->notificationsList->set($player, true);
        NotificationListCache::add($player);
        return $this->notificationsList->save() ? Promise::resolved() : Promise::rejected();
    }

    public function disablePlayerNotifications(string $player): Promise {
        $this->notificationsList->remove($player);
        NotificationListCache::remove($player);
        return $this->notificationsList->save() ? Promise::resolved() : Promise::rejected();
    }

    public function hasNotificationsEnabled(string $player): Promise {
        return Promise::resolved($this->notificationsList->get($player, false));
    }

    public function getNotificationList(): Promise {
        return Promise::resolved($this->notificationsList->getAll());
    }

    public function addToWhitelist(string $player): Promise {
        $this->maintenanceList->set($player, true);
        MaintenanceListCache::add($player);
        return $this->maintenanceList->save() ? Promise::resolved() : Promise::rejected();
    }

    public function removeFromWhitelist(string $player): Promise {
        $this->maintenanceList->remove($player);
        MaintenanceListCache::remove($player);
        return $this->maintenanceList->save() ? Promise::resolved() : Promise::rejected();
    }

    public function isOnWhitelist(string $player): Promise {
        return Promise::resolved($this->notificationsList->get($player, false));
    }

    public function getWhitelist(): Promise {
        return Promise::resolved(array_filter($this->maintenanceList->getAll(true), fn(string $user) => $this->maintenanceList->get($user, false)));
    }

    public function getTemplatesConfig(): ?Config {
        return $this->templatesConfig;
    }

    public function getServerGroupsConfig(): Config {
        return $this->serverGroupsConfig;
    }

    public function getModulesConfig(): Config {
        return $this->modulesConfig;
    }

    public function getNotificationsList(): Config {
        return $this->notificationsList;
    }

    public function getMaintenanceList(): Config {
        return $this->maintenanceList;
    }
}

File: src/main/java/de/pocketcloud/cloud/provider/CloudJsonProvider.java
package de.pocketcloud.cloud.provider;

import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.cache.ActiveInGameModuleCache;
import de.pocketcloud.cloud.cache.LocalCache;
import de.pocketcloud.cloud.cache.NotificationListCache;
import de.pocketcloud.cloud.cache.WhitelistCache;
import de.pocketcloud.cloud.config.Config;
import de.pocketcloud.cloud.config.exception.UnsupportedFileExtensionException;
import de.pocketcloud.cloud.config.type.JsonConfigType;
import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.template.Template;
import de.pocketcloud.cloud.template.group.ServerGroup;
import de.pocketcloud.cloud.util.PocketCloudPaths;
import de.pocketcloud.cloud.util.concurrent.Promise;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

public final class CloudJsonProvider extends CloudProvider {

    private Config templatesConfig;
    private Config serverGroupsConfig;
    private Config modulesConfig;
    private Config notificationsList;
    private Config maintenanceList;

    public CloudJsonProvider() {
        try {
            this.templatesConfig = new Config(PocketCloudPaths.templates().with("templates.json").asPath());
            this.serverGroupsConfig = new Config(PocketCloudPaths.groups().with("groups.json").asPath());
            this.modulesConfig = new Config(PocketCloudPaths.storage().inGame().with("modules.json").asPath(), new JsonConfigType(), Map.of(
                    ActiveInGameModuleCache.SIGN_MODULE, false,
                    ActiveInGameModuleCache.NPC_MODULE, false,
                    ActiveInGameModuleCache.HUB_COMMAND_MODULE, false
            ));

            this.notificationsList = new Config(PocketCloudPaths.storage().inGame().with("notificationList.json").asPath());
            this.maintenanceList = new Config(PocketCloudPaths.storage().inGame().with("maintenanceList.json").asPath());

            maintenanceList.getAll().keySet().stream().filter(p -> maintenanceList.get(p, false)).forEach(p -> LocalCache.get(WhitelistCache.class).add(p));
            notificationsList.getAll().keySet().stream().filter(p -> notificationsList.get(p, false)).forEach(p -> LocalCache.get(NotificationListCache.class).add(p));
            modulesConfig.getAll().keySet().stream().filter(m -> modulesConfig.get(m, false)).forEach(m -> LocalCache.get(ActiveInGameModuleCache.class).add(m));
        } catch (IOException | UnsupportedFileExtensionException e) {
            CloudLogger.get().exception("Unable to load CloudJsonProvider", e);
            PocketCloud.instance().shutdown();
        }
    }
    
    @Override
    public Promise<Void> addTemplate(Template template) {
        templatesConfig.set(template.name(), template.write());
        return saveConfig(templatesConfig);
    }

    @Override
    public Promise<Void> removeTemplate(Template template) {
        templatesConfig.remove(template.name());
        return saveConfig(templatesConfig);
    }

    @Override
    public Promise<Void> editTemplate(Template template, Map<String, Object> newData) {
        templatesConfig.set(template.name(), newData);
        return saveConfig(templatesConfig);
    }

    @Override
    public Promise<Optional<Template>> getTemplate(String templateName) {
        @SuppressWarnings("unchecked")
        Map<String, Object> data = templatesConfig.get(templateName, Map.class);

        try {
            if (data == null) return Promise.resolved(Optional.empty());
            Template template = Template.read(data);
            return Promise.resolved(Optional.of(template));
        } catch (Exception e) {
            return Promise.failed(e);
        }
    }

    @Override
    public Promise<Boolean> checkTemplate(String template) {
        return Promise.resolved(templatesConfig.get(template, false));
    }

    @Override
    public Promise<Map<String, Template>> getTemplates() {
        Map<String, Template> templates = new HashMap<>();
        templatesConfig.getAll().forEach((_, data) -> {
            try {
                @SuppressWarnings("unchecked")
                Template template = Template.read((Map<String, Object>) data);
                templates.put(template.name(), template);
            } catch (Exception e) {
                CloudLogger.get().exception("Unable to load templates", e);
            }
        });
        return Promise.resolved(templates);
    }
    
    @Override
    public Promise<Void> addServerGroup(ServerGroup serverGroup) {
        serverGroupsConfig.set(serverGroup.name(), serverGroup.write());
        return saveConfig(serverGroupsConfig);
    }

    @Override
    public Promise<Void> removeServerGroup(ServerGroup serverGroup) {
        serverGroupsConfig.remove(serverGroup.name());
        return saveConfig(serverGroupsConfig);
    }

    @Override
    public Promise<Void> editServerGroup(ServerGroup serverGroup, Map<String, Object> newData) {
        serverGroupsConfig.set(serverGroup.name(), newData);
        return saveConfig(serverGroupsConfig);
    }

    @Override
    public Promise<Optional<ServerGroup>> getServerGroup(String serverGroupName) {
        @SuppressWarnings("unchecked")
        Map<String, Object> data = serverGroupsConfig.get(serverGroupName, Map.class, null);

        try {
            if (data == null) return Promise.resolved(Optional.empty());
            ServerGroup group = ServerGroup.read(data);
            return Promise.resolved(Optional.of(group));
        } catch (Exception e) {
            return Promise.failed(e);
        }
    }

    @Override
    public Promise<Boolean> checkServerGroup(String serverGroup) {
        return Promise.resolved(serverGroupsConfig.has(serverGroup));
    }

    @Override
    public Promise<Map<String, ServerGroup>> getServerGroups() {
        Map<String, ServerGroup> serverGroups = new HashMap<>();
        serverGroupsConfig.getAll().forEach((_, data) -> {
            try {
                @SuppressWarnings("unchecked")
                ServerGroup group = ServerGroup.read((Map<String, Object>) data);
                serverGroups.put(group.name(), group);
            } catch (Exception _) {}
        });
        return Promise.resolved(serverGroups);
    }

    @Override
    public Promise<Void> setModuleState(String module, boolean enabled) {
        modulesConfig.set(module, enabled);
        LocalCache.get(ActiveInGameModuleCache.class).set(module, enabled);
        return saveConfig(modulesConfig);
    }

    @Override
    public Promise<Optional<Boolean>> getModuleState(String module) {
        return Promise.resolved(Optional.ofNullable(modulesConfig.get(module, null)));
    }

    @Override
    public Promise<Void> enablePlayerNotifications(String player) {
        notificationsList.set(player, true);
        LocalCache.get(NotificationListCache.class).add(player);
        return saveConfig(notificationsList);
    }

    @Override
    public Promise<Void> disablePlayerNotifications(String player) {
        notificationsList.remove(player);
        LocalCache.get(NotificationListCache.class).remove(player);
        return saveConfig(notificationsList);
    }

    @Override
    public Promise<Boolean> hasNotificationsEnabled(String player) {
        return Promise.resolved(notificationsList.get(player, false));
    }

    @Override
    public Promise<List<String>> getNotificationList() {
        return Promise.resolved(new ArrayList<>(notificationsList.getKeys()));
    }

    @Override
    public Promise<Void> addToWhitelist(String player) {
        maintenanceList.set(player, true);
        LocalCache.get(WhitelistCache.class).add(player);
        return saveConfig(notificationsList);
    }

    @Override
    public Promise<Void> removeFromWhitelist(String player) {
        maintenanceList.remove(player);
        LocalCache.get(WhitelistCache.class).remove(player);
        return saveConfig(notificationsList);
    }

    @Override
    public Promise<Boolean> isOnWhitelist(String player) {
        return Promise.resolved(maintenanceList.get(player, false));
    }

    @Override
    public Promise<List<String>> getWhitelist() {
        return Promise.resolved(maintenanceList.getAll().entrySet().stream()
                .filter(e -> Boolean.TRUE.equals(e.getValue()))
                .map(Map.Entry::getKey)
                .collect(Collectors.toList()));
    }

    private Promise<Void> saveConfig(Config config) {
        try {
            config.save();
            return Promise.resolved(null);
        } catch (Exception e) {
            return Promise.failed(e);
        }
    }
}